src/Avalonia.Base/Media/TextFormatting/CharacterBufferReference.cs

```diff
@@ -0,0 +1,176 @@
+        }
+
+        /// <summary>
+        /// Compute hash code
+        /// </summary>
+        public override int GetHashCode()
+        {
+            return CharacterBuffer.IsEmpty ? 0 : CharacterBuffer.GetHashCode();
+        }
+
+        /// <summary>
+        /// Test equality with the input object 
+        /// </summary>
+        /// <param name="obj"> The object to test. </param>
+        public override bool Equals(object? obj)
+        {
+            if (obj is CharacterBufferReference reference)
+            {
+                return Equals(reference);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Test equality with the input CharacterBufferReference
+        /// </summary>
+        /// <param name="value"> The characterBufferReference value to test </param>
+        public bool Equals(CharacterBufferReference value)
+        {
+            return CharacterBuffer.Equals(value.CharacterBuffer);
+        }
+
+        /// <summary>
+        /// Compare two CharacterBufferReference for equality
+        /// </summary>
+        /// <param name="left">left operand</param>
+        /// <param name="right">right operand</param>
+        /// <returns>whether or not two operands are equal</returns>
+        public static bool operator ==(CharacterBufferReference left, CharacterBufferReference right)
+        {
+            return left.Equals(right);
+        }
+
+        /// <summary>
+        /// Compare two CharacterBufferReference for inequality
+        /// </summary>
+        /// <param name="left">left operand</param>
+        /// <param name="right">right operand</param>
+        /// <returns>whether or not two operands are equal</returns>
+        public static bool operator !=(CharacterBufferReference left, CharacterBufferReference right)
+        {
+            return !(left == right);
+        }
+
+        public ReadOnlyMemory<char> CharacterBuffer { get; }
+
+        public int OffsetToFirstChar { get; }
+
+        /// <summary>
+        /// A MemoryManager over a raw pointer
+        /// </summary>
+        /// <remarks>The pointer is assumed to be fully unmanaged, or externally pinned - no attempt will be made to pin this data</remarks>
+        public sealed unsafe class UnmanagedMemoryManager<T> : MemoryManager<T>
+            where T : unmanaged
+        {
+            private readonly T* _pointer;
+            private readonly int _length;
+
+            /// <summary>
+            /// Create a new UnmanagedMemoryManager instance at the given pointer and size
+            /// </summary>
+            /// <remarks>It is assumed that the span provided is already unmanaged or externally pinned</remarks>
+            public UnmanagedMemoryManager(Span<T> span)
+            {
+                fixed (T* ptr = &MemoryMarshal.GetReference(span))
+                {
+                    _pointer = ptr;
+                    _length = span.Length;
+                }
```

Comment:
I am not sure, is it safe? It gets fixed/pinned pointer to the span, but after "fixed" statement nothing will prevent span to be removed from stack and underlying buffer to be moved by GC.
Also, I am not sure if it's used anywhere, unlikely UnmanagedMemoryManager(T* pointer, length). So can it be just removed for safity?

Reply:
With `UnmanagedMemoryManager(T* pointer, length)` caller is responsible to keep underlying buffer from being moved, so it's not a problem in this class.

Reply:
I think we can remove this wrapped and let just allow to create a buffer from ReadOnlyMemory

Reply:
So we don't have to maintain such wrapper